DA/Internal/Desugar.hs

```diff
@@ -152,3 +153,6 @@ newtype Method t i (m : Symbol) r = Method { unMethod : t -> r }
 
 mkMethod : (Implements t i, HasMethod i m r) => (t -> r) -> Method t i m r
 mkMethod = Method
+
+class Implements t i => HasExerciseInterface i t c r | i t c -> r where
```

Comment:
just for my own understanding: Do we need the superclass constraint here or would things work without it as well? I would have though the latter but maybe I’m missing something.

Reply:
No, it's not necessary, I added it because it "felt" like the right thing, and because it was in the context of `exerciseInterface` in the issue (https://github.com/digital-asset/daml/issues/11982). But yeah we can drop it if you want, nothing relies on that superclass.

Reply:
I’d lean towards not having it, seems to make it easier to evolve the typeclass structure in a backwards compatible way.

@sofiafaro-da wdyt?

Reply:
I agree that `Implements t i` shouldn't be in the typeclass. Actually I would prefer that we don't even include the `i` parameter in the typeclass here, and we have a class that's more like this:

```
class HasExerciseGuarded t c r | t c -> r where
    exerciseGuarded : (t -> Bool) -> ContractId t -> c -> Update r
```

This should make type inference work in our favor (don't need to specify the interface type `i`). I also have the sense that we might one day want to extend "guarded exercises" to template choices, which this would enable.

To define the guard `(i -> Bool)` that you need to pass for UExerciseInterface you would use something like `\x -> fromMaybe False p (fromInterface x)` where `p` is the `(t -> Bool)` argument. This can be done with a helper in DA.Internal.Desugar

Reply:
thanks for your suggestion @sofiafaro-da! I went ahead and implemented it as suggested :)